### PR TITLE
Implemented the take event method

### PR DESCRIPTION
**Pull request by dabonnie**
Added handling of the take event in both static publisher and static subscriber

Action items:

- which return codes should be used for each event?
- how to handle void * assignment properly?
- how to assign struct members from the DDS implementation layer to the RMW layer properly?

### CODE REVIEW

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp [hunk 1]
```diff
@@ -54,4 +54,84 @@ class ConnextPublisherListener : public DDS::PublisherListener
...
+        const DDS_StatusMask mask,
+        void * event) {
+
+  switch(mask) {
+
+    case DDS_LIVELINESS_LOST:
+
+      LivelinessLostStatus * liveliness_lost = topic_writer_->liveliness_lost_status();
+
+      rmw_liveliness_lost_t rmw_liveliness_lost = new liveliness_lost_t();
```
Comment (emersonknapp) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: This doesn't compile, does it? (and all same uses of `new` below) You are allocating new heap memory for a stack object

File: rmw_connext_shared_cpp/src/event.cpp [hunk 2]
```diff
@@ -49,5 +49,8 @@ __rmw_take_event(
     rmw_ret_code ret_code = custom_event_info.get_status(status_mask, event);
 
     //if ret_code is not okay, return error and set taken to false.
+    if(ret_code != RMW_RET_OK) {
```
Comment (mm318) [inline rmw_connext_shared_cpp/src/event.cpp]: Isn't `ret_code` and `taken` duplicate information here?

Comment (dabonnie) [reply rmw_connext_shared_cpp/src/event.cpp]: It seems that way, but I'm not sure how the defined signature affects the users. Need @nburek or @ross-desmond to weigh in.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp [hunk 3]
```diff
@@ -54,4 +54,84 @@ class ConnextPublisherListener : public DDS::PublisherListener
   std::atomic<std::size_t> current_count_;
 };
 
+/**
+ * Remap the specific RTI Connext DDS DataWriter Status to a generic RMW status type.
+ *
+ * @param mask input status mask
+ * @param event
+ */
+inline rmw_ret_code ConnextStaticSubscriberInfo::get_status(
```
Comment (emersonknapp) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: Do you mean `rmw_ret_t`? https://github.com/aws-robotics/rmw/blob/master/rmw/include/rmw/ret_types.h#L25

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 1 above)
Comment (mm318) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: So here you do
```
rmw_liveliness_lost_t * rmw_liveliness_lost = static_cast<rmw_liveliness_lost_t*>(event)
```
and then copy `liveliness_lost.total_count()` into `rmw_liveliness_lost->total_count`, etc.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 3 above)
Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: That would make sense. When pair programmed the return code was the typedef rmw_ret_code. Frustratingly enough I'm lacking context to make a useful comment about the design.

**Comment by mm318** [comment]
Great! I think this change is largely correct, just that you need to put the information from `topic_writer_` into the location that the `event` pointer is pointing to.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 3 above)
Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: I had to fix this in master branch of this fork as well. Probably just some oversight.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp [hunk 4]
```diff
@@ -54,4 +54,84 @@ class ConnextPublisherListener : public DDS::PublisherListener
...
+        const DDS_StatusMask mask,
+        void * event) {
+
+  switch(mask) {
```
Comment (mm318) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: This mask is obtained from `get_mask_from_rmw()`, so it is guaranteed to be one-hot (which is good!).

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp [hunk 5]
```diff
@@ -54,4 +54,84 @@ class ConnextPublisherListener : public DDS::PublisherListener
...
+      //TODO
+      //rmw_publication_matched_t.last_subscription_handle = topic_writer_->last_subscription_handle();
+
+      //todo return value?
+      //todo assign to void *?
+      break;
+
+    default:
+      //todo is this really ok?
+      return RMW_RET_OK;
```
Comment (mm318) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: The default case would not be OK.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 1 above)
Comment (emersonknapp) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: Ok - so the general pattern for the `//todo assign to void*` lines it that `event` is being passed in to be filled, based on `mask` we know what type to cast it to, and assume it has been correctly allocated as such.

File: rmw_connext_shared_cpp/src/event.cpp (hunk 2 above)
Comment (mm318) [reply rmw_connext_shared_cpp/src/event.cpp]: I believe we're completely free to change the signature of this function, because this is all our/new stuff.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_subscriber_info.hpp [hunk 6]
```diff
@@ -57,27 +59,114 @@ class ConnextSubscriberListener : public DDS::SubscriberListener
...
+
+      //todo return value?
+      //todo assign to void *?
       break;
+
     default:
-      return RMW_RET_ERROR;
+      //todo is this really ok?
+      return RMW_RET_OK;
   }
```
Comment (emersonknapp) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_subscriber_info.hpp]: for all the `//todo return value`, I would have a catch-all `return OK` as the very last line of the function, I am guessing that the `default` case for the `switch` should stay `ERROR`, and if there are any places in the middle where something can go wrong, catch and return error at those places

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 5 above)
Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: The other cases would be OK.

File: rmw_connext_shared_cpp/src/event.cpp (hunk 2 above)
Comment (emersonknapp) [reply rmw_connext_shared_cpp/src/event.cpp]: It's not duplicate informateon, because `ret_code` is not returned to the calling context. You could make this shorter by saying `*taken = (ret_code == RMW_RET_OK)` as a one liner

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 5 above)
Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: What should we return then? if the default is not OK are we saying that the QoS setting is not supported in the RMW layer?

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 4 above)
Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: But isn't DDS_StatusMask specific to RTI? I don't see how this is general in the RMW layer.

File: rmw_connext_shared_cpp/src/event.cpp (hunk 2 above)
Comment (mm318) [reply rmw_connext_shared_cpp/src/event.cpp]: `taken` and `ret_code` both end up here https://github.com/mm318/rcl/blob/master/rcl/src/rcl/event.c#L117 as `taken` and `ret`

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 5 above)
Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: Yeah, it means something like that, that we don't support retrieving information on the type of QoS event that was requested.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 4 above)
Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: The `get_mask_from_rmw()` that I was talking about is implemented in `rmw_connext_shared_cpp`, so it is specific to RTI.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 5 above)
Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: We could return something like `RMW_RET_INVALID_ARGUMENT` or some new enum.

Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: @mm318 agreed that we need a new code. Maybe simply RMW_UNSUPPORTED_QOS? Not sure why the mask, set outside of the users control, would be an invalid argument for their call? Maybe just semantics but I think it would be clearer.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 1 above)
Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: Thanks for the explanation and the quick pair program.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp [hunk 7]
```diff
@@ -62,62 +71,87 @@ class ConnextPublisherListener : public DDS::PublisherListener
...
+inline rmw_ret_t ConnextStaticPublisherInfo::get_status(const DDS_StatusMask mask, void * event)
+{
+  switch (mask) {
+    case DDS_StatusKind::DDS_LIVELINESS_LOST_STATUS: {
+        DDS_LivelinessLostStatus liveliness_lost;
+        DDS_ReturnCode_t dds_return_code =
+          topic_writer_->get_liveliness_lost_status(liveliness_lost);
+
+        rmw_ret_t from_dds;
+        if (!check_dds_ret_code(dds_return_code, from_dds)) {
```
Comment (mm318) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: I think we should generally prefer pure functions over functions that modify an input parameter. So for example, here you can do:
```
rmw_ret_t from_dds = check_dds_ret_code(dds_return_code);
if (RMW_RET_OK != from_dds) {
   return from_dds;
}
```

Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: The reason the above lines take in a `liveliness_lost` input parameter object to modify is because `DDS_LivelinessLostStatus` is a large object that may be expensive to return (copy elision is not guaranteed until C++17) and also because they needed to return a status code somehow.

File: rmw_connext_shared_cpp/src/event_converter.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      {RMW_EVENT_LIVELINESS_LOST,            DDS_LIVELINESS_LOST_STATUS},
+      {RMW_EVENT_OFFERED_DEADLINE_MISSED,    DDS_OFFERED_DEADLINE_MISSED_STATUS},
+      {RMW_EVENT_OFFERED_INCOMPATIBLE_QOS,   DDS_OFFERED_INCOMPATIBLE_QOS_STATUS},
+      {RMW_EVENT_PUBLICATION_MATCHED,        DDS_PUBLICATION_MATCHED_STATUS}
+
+    }};
+}
+
+DDS_StatusMask get_mask_from_rmw(const rmw_event_type_t & event_t) {
+  return mask_map.at(event_t);
```
Comment (mm318) [inline rmw_connext_shared_cpp/src/event_converter.cpp]: Thanks for moving this into a .cpp file!

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 7 above)
Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: Ok sounds good. Any reason to map the RMW and DDS statuses in a std::unordered_map rather than using a switch statement? Right now I don't see the use but someone else may have one.

File: rmw_connext_shared_cpp/src/event_converter.cpp (hunk 8 above)
Comment (dabonnie) [reply rmw_connext_shared_cpp/src/event_converter.cpp]: @emersonknapp's suggestion in an in-person code review :-)

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 7 above)
Comment (emersonknapp) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: > Any reason to map the RMW and DDS statuses in a std::unordered_map rather than using a switch statement?

Don't think it matters either way.

> pure functions over functions that modify an input parameter

@mm318 I think this is an imprecise distinction - it's a C-style version of returning a pair, which is a purely functional pattern. The more important point is that the code is more single-purpose and understandable (like you said) as

```
if (dds_ret_to_rmw_ret(dds_return_code) != RMW_RET_OK) {
return from_dds;
}
```

plus, that integer types like enums should just about always be used as

Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: Actually, I think the switch statement would be better, because it would be lighter on resources.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp [hunk 9]
```diff
@@ -17,25 +17,36 @@
...
 class ConnextPublisherListener;
 
 extern "C"
 {
-struct ConnextStaticPublisherInfo
+struct ConnextStaticPublisherInfo : ConnextCustomEventInfo
 {
   DDS::Publisher * dds_publisher_;
   ConnextPublisherListener * listener_;
-  DDS::DataWriter * topic_writer_;
```
Comment (emersonknapp) [inline rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: why did we get rid of these namespaces? Did a `using` sneak in somewhere?

File: rmw_connext_shared_cpp/src/event.cpp [hunk 10]
```diff
@@ -16,39 +16,46 @@
...
 #include "rmw_connext_shared_cpp/event.hpp"
 #include "rmw_connext_shared_cpp/event_converter.hpp"
 #include "rmw_connext_shared_cpp/types.hpp"
-#include "rmw_connext_shared_cpp/connext_static_event_info.hpp"
-
 
 rmw_ret_t
 __rmw_take_event(
-  const char * implementation_identifier,
+	const char * implementation_identifier,
```
[the remark this answers is not in the document]
Comment (dabonnie) [reply rmw_connext_shared_cpp/src/event.cpp]: Nope, I'm setup to use spaces.

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 9 above)
Comment (dabonnie) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: This shouldn't have been removed. Concerning it compiled.

File: rmw_connext_shared_cpp/src/event.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (emersonknapp) [reply rmw_connext_shared_cpp/src/event.cpp]: weird maybe it's just github's viewer

File: rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp (hunk 7 above)
Comment (mm318) [reply rmw_connext_cpp/include/rmw_connext_cpp/connext_static_publisher_info.hpp]: @emersonknapp I think if you're going to return a tuple, it would be more understandable if all elements of the tuple go in as a modifiable inputs and the return value is the status code, and the meaning of the status code should be orthogonal to the meaning of the elements in the tuple (as in the case of `get_liveliness_lost_status(liveliness_lost)`).

In this case, both the return value and the modifiable input pretty much mean the same thing, the code is basically too wordy.

File: rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp [hunk 11]
```diff
@@ -15,29 +15,14 @@
 #ifndef RMW_CONNEXT_SHARED_CPP__EVENT_CONVERTER_HPP_
 #define RMW_CONNEXT_SHARED_CPP__EVENT_CONVERTER_HPP_
 
-#include <unordered_map>
-
 #include "ndds/ndds_cpp.h"
-
+#include "ndds/ndds_namespace_cpp.h"
+#include "rmw_connext_shared_cpp/ndds_include.hpp"
+#include <unordered_map>
```
[the remark this answers is not in the document]
Comment (dabonnie) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp]: Good catch. How did you determine? I didn't get any warnings from building / testing / IDE.

[the remark this answers is not in the document]
Comment (emersonknapp) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp]: just noticed

[the remark this answers is not in the document]
Comment (mm318) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp]: I wonder if the linters that OSRF uses can catch unused `#include`s.

[the remark this answers is not in the document]
Comment (dabonnie) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp]: Not in their current test suite, but it looks like cppclean (https://stackoverflow.com/questions/614794/detecting-superfluous-includes-in-c-c) could be a possible solution. I might open a task to put in the backlog for investigation.

[the remark this answers is not in the document]
Comment (emersonknapp) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp]: https://github.com/ament/ament_lint/issues/64

[the remark this answers is not in the document]
Comment (dabonnie) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/event_converter.hpp]: I don't understand why they closed the issue, as I don't see how ament includes cppclean. Would be nice if it was supported out of the box.
